docs/root/overview.md

```diff
@@ -0,0 +1,189 @@
+**Pool** concepts. For executing requests, the pool will be requested to create
+a **StreamEncoder**, and Nighthawk will pass its own **StreamDecoderImpl** into
+that as an argument. The integration surface between **BenchmarkClient** is
+defined via `BenchmarkClient::tryStartRequest()` and a callback specification
+which will be fired upon completion of a successfully started request.
+
+For H/3, it is anticipated that it will fit into this model, but if all else
+fails, it will be entirely possible to wire in a new type of
+**BenchmarkClient**.
+
+### RequestSource
+
+**RequestSource** is an abstraction that allows us to implement different ways
+for **BenchmarkClient** to get information on what the request that it is about
+to fire off should look like. Today, two implementations exist: a static one,
+which will repeat the same request over and over, as well as one that pulls
+dynamic request data from a grpc service. The latter can, for example, be used
+to implement log-replay.
+
+### StreamDecoder
+
+**StreamDecoder** is a Nighthawk-specific implementation of an [Envoy
+concept](https://github.com/envoyproxy/envoy/blob/3156229006a5340b65c773329070737f67e81826/include/envoy/http/filter.h#L463).
+StreamDecoder will by notified by Envoy as headers and body fragments arrive.
+The Nighthawk implementation of that is responsible for coordinating lifetime
+events of a request to upper abstraction layers (**BenchmarkClient**,
+**Sequencer**) as well as recording latency and reporting that upwards.
+
+### OutputCollector
+
+**OutputCollector** is a container that facilitates building up the native output
+format of Nighthawk (`proto3`, `nighthawk::client::Output`). It is the basis for all
+output formats offered by Nighthawk, including CLI human output.
+
+### OutputFormatter
+
+**OutputFormatter** is responsible for transformations of nighthawk::client::Output
+to requested formats (e.g. human, json, fortio, etc)
+
+### Statistic
+
+Nighthawk’s **Statistic** is responsible for administrating latencies. The most
+notable implementation that exists today wraps
+[HdrHistogram](https://github.com/HdrHistogram/HdrHistogram_c), but Nighthawk
+also has a couple of other implementations which mostly exist to ensure that
+floating point math is correct in tests, as well as a simple efficient
+implementation that simply tracks the `mean` and `pstddev` for those cases where
+we don't need percentiles. For various reasons, HdrHistogram might get replaced
+by [libcirclhist](https://github.com/envoyproxy/nighthawk/issues/115) in the
+near future.
+
+### H1Pool & H2Pool
+
+Nighthawk derives its own version of these from the vanilla Envoy ones. It does
+that to implement things like pro-active connection pre-fetching and H2
+multi-connection support, as well as offer more connection management
+strategies.
+
+## Nighthawk binaries
+
+### nighthawk_client
+
+The CLI interface of the Nighthawk client. It will synthesize traffic according
```

Comment:
(nit) This seems to be the only place where we used the future tense, maybe:

The CLI interface of the Nighthawk client. It synthesizes...